src/main/java/net/sf/jsqlparser/parser/CCJSqlParserUtil.java

```diff
@@ -54,13 +63,25 @@ public static Statement parse(String sql) throws JSQLParserException {
      * @throws JSQLParserException
      */
     public static Statement parse(String sql, Consumer<CCJSqlParser> consumer) throws JSQLParserException {
-        boolean allowComplexParsing = getNestingDepth(sql)<=ALLOWED_NESTING_DEPTH;
-        
-        CCJSqlParser parser = newParser(sql).withAllowComplexParsing(allowComplexParsing);
-        if (consumer != null) {
-            consumer.accept(parser);
+        Statement statement = null;
+
+        // first, try to parse fast and simple
+        try {
+            CCJSqlParser parser = newParser(sql).withAllowComplexParsing(false);
```

Comment:
If complex parsing is initially switched off, it should not be switched on in this method. Right? However, this is only some kind of hack. The grammar itself is flawed if we need to do something like this. The porblem is the excessive use of Lookups, but if the grammar would be designed that a fixed lookup would be enough, then we do not need a switch like this.

Reply:
You are correct in your observation, but I respectfully do not share your conclusion:

1) It is not a `hack` but actually a completely valid approach (to my knowledge the only available method) to implement a `IF ... THEN ...` condition in the parser

2) While the grammar certainly could be rewritten/optimized, I still claim that SQL is a bit exceptional:
a) we try to be RDBMS/Dialect agnostic and so have to deal with a great flexibility in the Grammar
b) SQL lacks reliable `terminated Blocks`, which makes it difficult to define `Stop` tokens. What do I mean with this: Obviously everything inside brackets `(` `)` is a block and the closing brackets `)` terminates this block and we know that we can stop parsing an expression at the closing bracket. Unfortunately SQL was "designed" for assistant secretaries, who were supposed to query data without programming skills.

Best illustration is issue #1444:
```sql
select doc.fullName from XWikiDocument doc where doc.fullName like :fullName escape '/' and doc.author like :author escape '!'
```

Is this
```sql
select doc.fullName from XWikiDocument doc where ( doc.fullName like :fullName escape '/' ) 
and ( doc.author like :author escape '!' )
```

Or is this
```sql
select doc.fullName from XWikiDocument doc where doc.fullName like :fullName escape ( '/' 
and doc.author like :author escape '!' )
```

JavaCC by its nature lacks a lot of functionality to deal with those requirements. JavaCC 21 added some features for a reason and I believe ANTLR is there for a reason too.

Long story short:
Yes, you are right: too many `lookup` harms us badly. But I am also right: we do need those  for parsing certain valid statements (look how functionality has grown since 4.0 and how the "new issue" rate went down since!)

So the suggest approach, while not a beauty at all solves that in a pragmatic way:
It tries to parse first without those slow `lookup` and only when this fails, it would activate the `lookup` -- because they are needed.

Reply:
Of course, if anyone can point me on how to avoid those `lookup` without losing functionality, I am all in!